Zend/zend_call_stack.h

```diff
@@ -0,0 +1,96 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Arnaud Le Blanc <arnaud.lb@gmail.com>                       |
+   +----------------------------------------------------------------------+
+*/
+
+#ifndef ZEND_CALL_STACK_H
+#define ZEND_CALL_STACK_H
+
+#include "zend.h"
+#include "zend_portability.h"
+#ifdef __APPLE__
+# include <pthread.h>
+#endif
+
+typedef struct _zend_call_stack {
+	void *base;
+	size_t max_size;
+} zend_call_stack;
+
+ZEND_API void zend_call_stack_init(void);
+
+ZEND_API bool zend_call_stack_get(zend_call_stack *stack);
+
+/** Returns an approximation of the current stack position */
+static zend_always_inline void *zend_call_stack_position(void) {
+#ifdef ZEND_WIN32
+	return _AddressOfReturnAddress();
+#elif HAVE_BUILTIN_FRAME_ADDRESS
+	return __builtin_frame_address(0);
+#else
+	void *a;
+	void *pos = (void*)&a;
+	return pos;
+#endif
+}
+
+static zend_always_inline bool zend_call_stack_overflowed(void *stack_limit) {
+	return (uintptr_t) zend_call_stack_position() <= (uintptr_t) stack_limit;
+}
+
+static inline void* zend_call_stack_limit(void *base, size_t size, size_t reserved_size)
+{
+	if (UNEXPECTED(size > (uintptr_t)base)) {
+		return (void*)0;
+	}
+
+	base = (int8_t*)base - size;
+
+	if (UNEXPECTED(UINTPTR_MAX - (uintptr_t)base < reserved_size)) {
+		return (void*)UINTPTR_MAX;
+	}
+
+	return (int8_t*)base + reserved_size;
+}
+
+static inline size_t zend_call_stack_default_size(void)
+{
+#ifdef __linux__
+	return 8 * 1024 * 1024;
+#endif
+#ifdef __FreeBSD__
+	return 8 * 1024 * 1024;
+#endif
+#ifdef ZEND_WIN32
+	// General
+	// https://docs.microsoft.com/en-us/windows/win32/procthread/thread-stack-size
+	return 1 * 1024 * 1024;
+	// IIS
+	// https://docs.microsoft.com/en-us/troubleshoot/developer/webapps/iis/site-behavior-performance/default-maximum-stack-size-thread
+	return 256 * 1024;
```

Comment:
These values do not match our custom stack size of 0x400000 (i.e. 64MiB):

https://github.com/php/php-src/blob/d952419a2ba9096a5d99dfe9acbac7ffc6d353eb/sapi/cli/config.w32#L10

and 0x800000 (i.e. 8MiB):

https://github.com/php/php-src/blob/d952419a2ba9096a5d99dfe9acbac7ffc6d353eb/sapi/phpdbg/config.w32#L19

(it looks like the phpdbg stack size has just been overlooked when [increasing the stack sizes](https://github.com/php/php-src/commit/54906c760fb578079b6783a2aa4184372a3179d3))

Anyhow, I don't think we should hard-code any value here, but leave that for the configuration.

Reply:
I noticed the custom stack size for the CLI (I did not notice the phpdbg one), but I was trying to find a common bound that would be low enough in all SAPIs. (Although the bogus double return breaks this.)

These hard coded values are used as a fallback for when auto detection of the stack size fails, when `zend.max_allowed_stack_size` is set to "auto detect" (the default). An alternative would be to fallback to "no stack limit".

On Windows I don't think that the auto detection can fail, so I could at least remove the Windows case here.

Reply:
> I was trying to find a common bound that would be low enough in all SAPIs.

Probably no chance, since besides users could do custom builds, apparently under Apache 2.4.39 the default stack size is only 0x40000 (i.e. 256KiB), although that can be changed. And the info about IIS is not relevant for PHP, since we do not support NSAPI anymore, but only FCGI.

Anyhow, thanks for the explanation of the hard-coded values!

> On Windows I don't think that the auto detection can fail, so I could at least remove the Windows case here.

Yeah, that might be the best solution; if there are issues, we could still introduce some fallback value.